crates/typst-cli/src/download.rs

```diff
@@ -0,0 +1,204 @@
+use std::collections::VecDeque;
+use std::io::{self, ErrorKind, Read, Stdout, Write};
+use std::time::{Duration, Instant};
+
+use ureq::Response;
+
+/// Keep track of this many past download amounts
+const DOWNLOAD_TRACK_COUNT: usize = 5;
+
+/// A wrapper around [`ureq::Response`] that reads the response body in chunks
+/// over a websocket and displays statistics about its progress.
+///
+/// Downloads will _never_ fail due to statistics failing to print, print errors
+/// are silently ignored.
+///
+/// [`ureq::Response`]: https://docs.rs/ureq/2.7.1/ureq/struct.Response.html
+pub struct RemoteReader {
+    reader: Box<dyn Read + Send + Sync + 'static>,
+    content_len: usize,
+    total_downloaded: usize,
+    downloaded_this_sec: usize,
+    downloaded_last_few_secs: VecDeque<usize>,
+    start_time: Instant,
+    last_print: Option<Instant>,
+    displayed_charcount: Option<usize>,
+    terminal: Stdout,
```

Comment:
I wonder whether this should also print to stderr. Probably, yes: https://unix.stackexchange.com/questions/331611/do-progress-reports-logging-information-belong-on-stderr-or-stdout

Reply:
Yes, let's swap.